Snakefile

```diff
@@ -0,0 +1,402 @@
+import subprocess
+
+#################################################################
+####################### general setup ###########################
+#################################################################
+
+# certain files can be fetched from the database or downloaded from our S3 bucket.
+# config["fetch_from_database"] dictates this
+# We could implement this via remote snakemake inputs (such as in nextstrain/ncov),
+# conditional rules (a bit of an anti-pattern) or ruleorders.
+ruleorder: download_previous_biosample > fetch_biosample_from_database
+ruleorder: download_previous_raw_ndjson > fetch_from_database
+if config.get("fetch_from_database", False):
+    ruleorder: fetch_biosample_from_database > download_previous_biosample
+    ruleorder: fetch_from_database > download_previous_raw_ndjson
+
+send_notifications = "SLACK_CHANNELS" in os.environ and "SLACK_TOKEN" in os.environ
+
+#################################################################
+################ work out what steps to run #####################
+#################################################################
+
+def _compute_targets(wildcards):
+    database = config["database_name"]    
+    targets = [f"data/{database}/upload.done"]
+    if config.get("trigger_preprocessing", False):
+        targets.append(f"data/{database}/trigger-preprocessing.done")
+    if send_notifications:
+        targets.append(f"data/{database}/notify.done")
+    if config.get("fetch_from_database", False):
+        targets.append(f"data/{database}/raw.upload.done")
+    return targets
+
+rule all:
+    input: _compute_targets
+
+#################################################################
+###################### rule definitions #########################
+#################################################################
+
+
+def run_shell_command_n_times(cmd, msg, cleanup_failed_cmd, retry_num=5):
```

Comment:
It would be neat to replace this with [built-in rule-specific retry behaviour](https://github.com/snakemake/snakemake/issues/367), if that ever lands in Snakemake.

Reply:
Agreed! This is now documented for posterity in #239